util/semverutil/semverutil.go

```diff
@@ -0,0 +1,41 @@
+package semverutil
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Version is a semantic version number.
+type Version struct {
+	Major int
+	Minor int
+	Patch int
+	Tail  string
+}
+
+// Parse parses a semantic version number.
+func Parse(s string) (Version, error) {
+	s = strings.TrimPrefix(s, "v")
+	var v Version
+	n, err := fmt.Sscanf(s, "%d.%d.%d%s", &v.Major, &v.Minor, &v.Patch, &v.Tail)
+	if err == io.EOF && n == 3 { // no tail case
+		return v, nil
+	}
+	if err != nil {
+		return Version{}, errors.Wrap(err, "parsing semantic version")
+	}
+	return v, nil
+}
+
+// String returns the string representation of the version.
+func (v Version) String() string {
+	return fmt.Sprintf("v%d.%d.%d%s", v.Major, v.Minor, v.Patch, v.Tail)
+}
+
+// IsCompatible returns true if the two versions are compatible.
+func IsCompatible(a, b Version) bool {
+	return a.Major == b.Major && a.Minor == b.Minor
```

Comment:
based on the name `IsCompatible`, is your idea that we would eventually replace the logic in this function to list known pairs that are compatible?

buildkit already provides a capabilities feature under https://github.com/earthly/buildkit/blob/earthly-main/solver/pb/caps.go

We've extended this in the past to force the client to require the `CapExecMountSock` capability: e.g. https://github.com/earthly/buildkit/commit/df14b0779eabc92481915e8a87d88065253832ba

I worry that having two mechanisms for this will be confusing.

Reply:
True... Not sure what the right solution is.

I would like to show a warning of some sort if Earthly and Buildkit have drastically different versions. I think that today we don't have a good way to track every buildkit change through capabilities (e.g. changes to the debugger, or changes to the WITH DOCKER wrapper script), so I think that this is not a good use-case for capabilities.

I figured that using major + minor comparison would give us a rough idea about whether Earthly and Buildkit would be close in terms of version number (e.g. 1.2.3 is close to 1.2.5, but not to 1.3.3). I wasn't thinking of extending this with specific version pairs - just to do semver comparison for this warning message only, and not for detecting support for specific pieces of functionality (where I think capabilities are the way to go).

Reply:
If we don't want to expand this to suggest stable pairings, then this seems manageable.

I suspect we've abused our patch number where we really should have increased our minor version.

Ultimately if this will  just issue a warning if, for instance, someone runs 0.6 against 0.5, that seems fine.